### ISSUE TITLE
Streams with IsEqualTo()

### ISSUE THREAD

**Issue by SilNak** [tpierrain/NFluent#158]
Hi Folks,

Consider the following Test:

```
[TestMethod]
        public void Stream_Equality_Test()
        {
            using (MemoryStream sourceStream = new MemoryStream())
            {
                using (MemoryStream targetStream = new MemoryStream())
                {

                    // Populate source
                    Random rand = new Random();
                    for (int i = 0; i < 123456; i++)
                        sourceStream.WriteByte((Byte)rand.Next(Byte.MinValue, Byte.MaxValue));

                    // copy source to target
                    sourceStream.WriteTo(targetStream);

                    Check.That(sourceStream.Length).IsEqualTo(targetStream.Length);

                    ////////////////////////////////////////
                    // This succeed (it compares contents)
                    sourceStream.Seek(0, SeekOrigin.Begin);
                    targetStream.Seek(0, SeekOrigin.Begin);
                    while (sourceStream.Position != sourceStream.Length)
                        Check.That(sourceStream.ReadByte()).IsEqualTo(targetStream.ReadByte());

                    ////////////////////////////////////////
                    // This fails (it compares Hashcodes)
                    Check.That(sourceStream).IsEqualTo(targetStream);
                }
            }
        }

```

In my quest to check to streams" contents, **I expect the IsEqualTo() to work fine, but it visibly defaults to "Hashcode" comparison**...

```
The following error states this:
The checked value:
	[System.IO.MemoryStream] with HashCode: [54870941]
The expected value:
	[System.IO.MemoryStream] with HashCode: [47753423]

```

Any Idea or better to check streams' contents ?

**Comment by tpierrain** [comment]
Hi @SilNak 

Thanks for using NFluent. Its IsEqualTo() check relies on object.Equals(instance, expected). In your test, the 2 streams aren't equals. This is why NFluent tells you that they are differents. The 2 instances HashCodes provided by NFluent error message is added just to help you find out they are 2 different instances despite the fact that they are of the same type.

Can you please provide what is your goal or use case so that we can improve the lib?

**Comment by tpierrain** [comment]
Now I'm wondering... do you need something like:
```c#
     Check.That(sourceStream).HasSameContentAs(targetStream);
```
?

**Comment by tpierrain** [comment]
Hello @SilNak 

I just implemented HasSameContentAs check. It will be provided shortly in the next release.

https://github.com/tpierrain/NFluent/blob/master/NFluent.35.Tests/StreamTests.cs

**Comment by SilNak** [comment]
Hi tpierrain,

  Sorry to be confused with IsEqualTo. Effectively the behavior is correct. 
  For the need, HasSameContentAs is what desired. 
  Thank you for your reactivity!
  I now use NFluent extensively in projects, it is very reliable and smart to use. 

Regards.

**Comment by dupdob** [comment]
I have doubt regarding naming choice: HasSameContent.
Stream are more like iterators than arrays. They do not have content...

**Comment by tpierrain** [comment]
Ok. What do you think about:

     Check.That(stream1).HasSameSequenceOfBytesAs(stream2);

?

**Comment by SilNak** [comment]
HasSameSequenceOfBytesAs clearly announce what it does, by the way.

**Comment by dupdob** [comment]
It is definitely a better suited name. Maybe a tad long, maybe HasSameBytesSequenceAs ?
Or what about: AccessSameBytesAs ? (maybe 3 s in a row is hard to read)

Quick questions regarding the actual check? are the streams 'rewound' to start (using seeek) ?

**Comment by tpierrain** [comment]
Thanks for your proposal @dupdob 
Since checks starting with "Has.." Is somehow idiomatic with NFluent, I'd rather pick HasSameBytesSequencesAs. 
As you can check by yourself here : 

https://github.com/tpierrain/NFluent/blob/master/NFluent.35/Assertions/StreamCheckExtensions.cs

this check doesn't have side effect on the stream. 

I'm currently migrating NFluent tests towards the latest nunit version (huge amount of works since there is no more ExpectedException attribute ;-( in order to prepare the next release of NFluent (at last ;-). I'll get back to you just after my nunit migration (hopefully today).

**Comment by dupdob** [comment]
Regarding streams: warning, support for seek operation is optional. E.g: network stream do not support for seek operation

**Comment by dupdob** [comment]
For NUnit, I already made the appropriate migration effort. Have you checked my latest PR?

**Comment by tpierrain** [comment]
You're right. I'll add the finally block. What do you suggest for the non seekable stream? to add checks support only for stream concrete types that are compatible?

**Comment by dupdob** [comment]
The main problem is that the Stream abstraction is leaky regarding the seekable nature of the actual data storage/medium.
If we are talking storage, think file/memory, then the check must be performed as expected.
If we are talking streams, think socket/pipe, then it gets hairy: not seekable, no clear definition of when the stream ends.
As of now, the best approach would be to provide support only for seekable stream  by offering the adequate overloads of the test, but it means something like 20+ signature for that.
I need to think about an alternative

**Comment by tpierrain** [comment]
Yes, and I didn't see any dismantling of all ExpectedException attributes.
Is there anything else for me to check?
